consensus/polybft/slashing/double_signing_tracker.go

```diff
@@ -0,0 +1,218 @@
+	} else if senderMsgs, ok = roundMsgs[view.Round]; !ok {
+		senderMsgs = createSenderMsgsMap(view.Round, roundMsgs)
+	}
+
+	msgs := m.getSenderMsgsLocked(msg.View, sender)
+	if msgs == nil {
+		msgs = []*ibftProto.Message{msg}
+	} else {
+		msgs = append(msgs, msg)
+	}
+
+	senderMsgs[sender] = msgs
+}
+
+// createSenderMsgsMap initializes senders message map for the given round
+func createSenderMsgsMap(round uint64, roundMsgs map[uint64]SenderMessagesMap) SenderMessagesMap {
+	sendersMsgs := make(SenderMessagesMap)
+	roundMsgs[round] = sendersMsgs
+
+	return sendersMsgs
+}
+
+type DoubleSigningTracker interface {
+	Handle(msg *ibftProto.Message)
+	GetEvidences(height uint64) []*DoubleSignEvidence
+	PruneMsgsUntil(height uint64)
+}
+
+type DoubleSigningTrackerImpl struct {
+	preprepare  *Messages
+	prepare     *Messages
+	commit      *Messages
+	roundChange *Messages
+
+	logger hclog.Logger
+}
+
+func NewDoubleSigningTracker(logger hclog.Logger) *DoubleSigningTrackerImpl {
+	return &DoubleSigningTrackerImpl{
+		logger:      logger,
+		preprepare:  &Messages{content: make(MessagesMap)},
+		prepare:     &Messages{content: make(MessagesMap)},
+		commit:      &Messages{content: make(MessagesMap)},
+		roundChange: &Messages{content: make(MessagesMap)},
+	}
+}
+
+// Handle is implementation of IBFTMessageHandler interface, which handles IBFT consensus messages
+func (t *DoubleSigningTrackerImpl) Handle(msg *ibftProto.Message) {
+	if err := t.validateMsg(msg); err != nil {
+		t.logger.Debug("[ERROR] invalid IBFT message retrieved", "error", err, "message", msg)
+
+		return
+	}
+
+	sender := types.BytesToAddress(msg.From)
+
+	msgMap := t.resolveMessagesStorage(msg.GetType())
+	msgMap.addMessage(msg.View, sender, msg)
+}
+
+// PruneMsgsUntil deletes all messages maps until the specified height
+func (t *DoubleSigningTrackerImpl) PruneMsgsUntil(height uint64) {
+	for _, msgType := range ibftProto.MessageType_value {
+		msgs := t.resolveMessagesStorage(ibftProto.MessageType(msgType))
+		if msgs == nil {
+			continue
+		}
+
+		for msgHeight := range msgs.content {
+			if msgHeight < height {
+				delete(msgs.content, msgHeight)
+			}
+		}
+	}
+}
+
+// GetEvidences returns double signing evidences for the given height
+func (t *DoubleSigningTrackerImpl) GetEvidences(height uint64) []*DoubleSignEvidence {
+	var (
+		evidences = []*DoubleSignEvidence{}
+		evidence  *DoubleSignEvidence
+	)
+
+	for _, msgType := range ibftProto.MessageType_value {
+		msgs := t.resolveMessagesStorage(ibftProto.MessageType(msgType))
+		if msgs == nil {
+			continue
+		}
+
+		roundMsgs, ok := msgs.content[height]
+		if !ok {
+			continue
+		}
+
+		for round, senderMsgs := range roundMsgs {
+			for address, msgs := range senderMsgs {
+				if len(msgs) <= 1 {
+					continue
+				}
+
+				firstSig := msgs[0].Signature
+				for _, msg := range msgs[1:] {
+					if !bytes.Equal(firstSig, msg.Signature) {
+						if evidence == nil {
```

Comment:
This seems wrong. If we have two validators that double signed on the same height and round, wouldn't it accumulate messages of the second validator to the ones that were already collected for the first one? Since `evidence` is nil only at the beginning of the execution.

Reply:
Good catch. Fixed https://github.com/0xPolygon/polygon-edge/pull/1756/commits/f34d131a162e6fe45a6bcdbf5d0ebbf1b96ce58f